core/controllers/learner_dashboard.py

```diff
@@ -0,0 +1,206 @@
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for the learner dashboard."""
+
+from core.controllers import base
+from core.domain import collection_services
+from core.domain import config_domain
+from core.domain import exp_services
+from core.domain import learner_progress_services
+from core.domain import subscription_services
+from core.domain import summary_services
+from core.domain import user_services
+import feconf
+import utils
+
+class LearnerDashboardPage(base.BaseHandler):
+    """Page showing the user's learner dashboard."""
+
+    @base.require_user
+    def get(self):
+        if self.username in config_domain.BANNED_USERNAMES.value:
+            raise self.UnauthorizedUserException(
+                'You do not have the credentials to access this page.')
+        elif user_services.has_fully_registered(self.user_id):
+            self.values.update({
+                'nav_mode': feconf.NAV_MODE_LEARNER_DASHBOARD
+            })
+            self.render_template(
+                'pages/learner_dashboard/learner_dashboard.html',
+                redirect_url_on_logout='/')
+        else:
+            self.redirect(utils.set_url_query_parameter(
+                feconf.SIGNUP_URL, 'return_url', feconf.LEARNER_DASHBOARD_URL))
+
+
+class LearnerDashboardHandler(base.BaseHandler):
+    """Provides data for the user's learner dashboard page."""
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    def get(self):
+        """Handles GET requests."""
+        if self.user_id is None:
+            raise self.PageNotFoundException
+
+        incomplete_exploration_ids = (
```

Comment:
Moved this logic to the backend. Also, the learner dashboard page has `@base.require_user` so we do redirect the learners. The data handler controller right? This is modeled after the creator dashboard handler method.